ext/opcache/zend_file_cache.c

```diff
@@ -1653,13 +1752,29 @@ zend_persistent_script *zend_file_cache_script_load(zend_file_handle *file_handl
 	int cache_it = 1;
 	unsigned int actual_checksum;
 	int ok;
+	int is_opfile = 0;
 
 	if (!full_path) {
 		return NULL;
 	}
-	filename = zend_file_cache_get_bin_file_path(full_path);
 
-	fd = zend_file_cache_open(filename, O_RDONLY | O_BINARY);
+	if(ZCG(accel_directives).allow_direct_exec_opcode) {
```

Comment:
The current implementation may also leak information as-is. When opcache is disabled or when allow_direct_exec_opcode is disabled, I assume php will interpret the `<?phpo` as an ordinary file and usually compile and echo it (unless it runs into a syntax issue if `<?php` or `<?=` tokens occur in the file.
(or a syntax error if `short_open_tags` support is compiled and enabled, but that's rare)

So if there's a snippet like this in a file (standalone script, config file, etc) that was published to a web server for a reason such as minification or obfuscation

```php
<?php // assume this was converted to a binary <?phpo file, not the original source
$host = 'some-db-host.com';
$dbUser = 'myuser';
$password = 'passblahblahblah';
$conn = somedb_login($host, $dbUser, $password);
somedb_exec('select some_proc();');
```

And if opcache is disabled, uninstalled, or allow_direct_exec_opcode is later absent or disabled, then I assume the web server would usually send the opcodes like it would for ordinary inline html, and anyone seeing that could guess the credentials just by looking at the `strings` of the raw file, or by finding an open source opcode disassembler.

Changing the prefix to something like `<?php /*MAGIC*/ throw new Error('allow_direct_exec_opcode support missing');  __halt_compiler();$BINARY_DATA` could be used to avoid that

Reply:
if `allow_direct_exec_opcode` disable, opcode file just plain binary files. 
If checks and prompts are added, the first few characters of the comparison file will be read at any time, which will affect OPCache's current performance